appsrc/simple/ge9800/ge98id.man
.TH GE98ID 1 "05 April 1998" "DICOM PS3" "DICOM PS3 - GE 9800 dump identifiers"
.SH NAME
ge98id \- ACR/NEMA DICOM PS3 ... DICOM PS3 - GE 9800 dump identifiers
.SH SYNOPSIS
.HP 10
.B ge98id
.B < " infile"
.SH DESCRIPTION
.LP
.B ge98id
reads the GE 9800 CT input file and dumps identifying attributes.
.SH OPTIONS
The descriptions of the identifiers go to standard out.
Any error messages go to standard error.
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
\ 
.RE
% ge98id <test.YP
.RE
StudyID="8541"
.RE
PatientID="12345678"
.RE
PatientName="CITIZEN DOE"
.RE
InstitutionName="ST. ELSEWHERE"
.RE
StudyDescription="THIS IS A NON-STUDY DESCRIPTION"
.RE
SeriesNumber="1"
.RE
AcquisitionNumber="6"
.RE
ImageNumber="6"
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR ge98topgm(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/simple/ge9800/ge98id.cc
#include <iostream.h>

#include "basetype.h"

// The following includes and definitions mirror libsrc/include/dctool/fltype.h ...

#include <float.h>

#ifdef POWINTEGEREXPONENTTYPE
#define	powi(m,e)	pow(m,(POWINTEGEREXPONENTTYPE)(e))
#else
#define	powi(m,e)	pow(m,(long)(e))
#endif

static Uint16
read16(istream& instream)
{
	Uint16 u;
	unsigned char buffer[2];
	instream.read((char *)buffer,2);
	// DataGeneral is a Big Endian machine
	u =  (Uint16)buffer[0];
	u <<= 8;
	u |= (Uint16)buffer[1];
	return u;
}

static double
readReal(istream& instream)
{
	double value;
	typedef struct {
		unsigned	sign : 1;
		unsigned	exponent : 7;
		unsigned	mantissa : 24;
	} DG_FLOAT;
	
	DG_FLOAT number;

	//unsigned char buffer[4];
	char buffer[4];
	instream.read(buffer,4);
	// DataGeneral is a Big Endian machine
	memcpy ((char *)(&number),buffer,4);

	unsigned char	sign = number.sign;
	Uint16		exponent = number.exponent;
	Uint32		mantissa = number.mantissa;

	value = (double) mantissa / (1 << 24) * powi (16.0, (long)(exponent) - 64);
	value = (sign == 0) ? value : -value;
	return value;
}

static OurStreamPos
seekword(istream& instream,unsigned block,unsigned word)
{
	// 256 word blocks,16 bit words ...
	const unsigned offset = (block*256+word)*2;
	instream.seekg(offset,ios::beg);
	return instream.tellg();
}

int
main(int argc,char *argv[])
{
	if (argc != 1) {
		cerr << "Usage: " << argv[0] << " < infile\n" << flush;
		return 1;
	}

	(void)seekword(cin,0,34);
	unsigned examheaderblock=((unsigned)read16(cin)-1)/256;
	(void)seekword(cin,0,35);
	unsigned imageheaderblock=((unsigned)read16(cin)-1)/256;

#define doascii(b,o,s,n)				\
	{						\
		char tmp[s+1];				\
		(void)seekword(cin,b,o);		\
		cin.read(tmp,s);			\
		tmp[s]=0;				\
		cout << n << "=\"" << tmp << "\"\n";	\
	}

#define doword(b,o,n)					\
	{						\
		(void)seekword(cin,b,o);		\
		unsigned tmp=read16(cin);		\
		cout << n << "=\"" << tmp << "\"\n";	\
	}

#define doreal(b,o,n)					\
	{						\
		(void)seekword(cin,b,o);		\
		double tmp=readReal(cin);		\
		cout << n << "=\"" << tmp << "\"\n";	\
	}

	doword (examheaderblock ,3    ,"StudyID");		// Exam Number
	doascii(examheaderblock ,11,12,"PatientID");		// Patient ID
	doascii(examheaderblock ,17,30,"PatientName");		// Patient Name
	doascii(examheaderblock ,47,30,"InstitutionName");	// Hospital Name
	doascii(examheaderblock ,76,30,"StudyDescription");	// Exam Description

	doword (imageheaderblock,10   ,"SeriesNumber");		// Position (study) number
	doword (imageheaderblock,46   ,"AcquisitionNumber");	// Scan Number
	doword (imageheaderblock,47   ,"ImageNumber");		// Image Number

	doreal (imageheaderblock,88   ,"SliceThickness");	// Aperture (1.5,3,5,10) (fp)
	doword (imageheaderblock,90   ,"KVP");				// Generator KVP (80,120,140)
	doword (imageheaderblock,91   ,"TubeCurrent");		// Tube current (10-360)
	doreal (imageheaderblock,92   ,"GantryAngle");		// Gantry Angle 
	doreal (imageheaderblock,94   ,"TableHeight");		// Table Height (mm) (fp)
	doreal (imageheaderblock,96   ,"SlicePosition");	// Axial Table Location (fp)
	doreal (imageheaderblock,98   ,"ScanTime");			// Scan time (1.4-9.4) (fp) 
	doreal (imageheaderblock,215  ,"ReconPixelSize");	// actual image pixel size (mm,fp) (scout), standard pixel size/target factor (recon) 

	return 0;
}
